realm/realm-library/src/main/java/io/realm/RealmInteger.java

```diff
@@ -50,111 +51,70 @@ public RealmInteger(String value) {
      */
-    public void set(long newValue) {
-        value = newValue;
-    }
+    public abstract void set(long newValue);
 
     /**
      * Increments the {@code RealmInteger}, adding the value of the argument.
      * Increment/decrement from all devices are reflected in the new value, which is guaranteed to converge.
      *
      * @param inc quantity to be added to the counter.
      */
-    public void increment(long inc) {
-        value += inc;
-    }
+    public abstract void increment(long inc);
 
     /**
      * Decrements the {@code RealmInteger}, subtracting the value of the argument.
      * Increment/decrement from all devices are reflected in the new value, which is guaranteed to converge.
      *
      * @param dec quantity to be subtracted from the counter.
      */
-    public void decrement(long dec) {
-        value -= dec;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public byte byteValue() {
-        return (byte) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public double doubleValue() {
-        return (double) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public float floatValue() {
-        return (float) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public int intValue() {
-        return (int) value;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public long longValue() {
-        return value;
-    }
+    public abstract void decrement(long dec);
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public short shortValue() {
-        return (short) value;
+    public String toString() {
+        return String.valueOf(longValue());
     }
 
     /**
-     * {@inheritDoc}
+     * RealmIntegers compare strictly by their values.
+     * Final to ensure contract over all subclasses
+     *
+     * @param o the compare target
+     * @return -1, 0, or 1, depending on whether this object's value is &gt;, =, or &lt; the target's.
      */
     @Override
-    public int compareTo(RealmInteger o) {
-        long otherValue = o.value;
-        return (value == otherValue) ? 0 : ((value > otherValue) ? 1 : -1);
-    }
+    public final int compareTo(RealmInteger o) {
+        long otherValue = o.longValue();
+        long thisValue = longValue();
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public String toString() {
-        return String.valueOf(value);
+        return (thisValue == otherValue) ? 0 : ((thisValue > otherValue) ? 1 : -1);
     }
 
     /**
-     * {@inheritDoc}
+     * A RealmInteger's hash code depends only on its value.
+     * Must be final to ensure contract over all subclasses
+     *
+     * @return true if the target has the same value.
      */
     @Override
-    public int hashCode() {
-        return (int) (value ^ (value >>> 32));
+    public final int hashCode() {
+        long thisValue = longValue();
```

Comment:
when the managed `RealmInteger` becomes valid, `longValue()` will throw an exception. in this case, the `hashCode()` will throw. But it doesn't sounds like a user expected behaviour? Same for `equals()`. `equals()` would be simpler i think, any invalid `RealmInteger` should not equal to anything.